### PR TITLE
Improved testing of token classes.

### PR DESCRIPTION
**Pull request by CH-JonathanBlake**
__Short description outlining key changes/additions__

Refactoring of Tokens in preparation to be moved into Oauth library

__JIRA Ticket Number__

https://companieshouse.atlassian.net/browse/FA-748

### Type of change

* [ ] Bug fix
* [ ] New feature
* [ ] Breaking change

### Pull request checklist

* [X] I have added unit tests for new code that I have added
* [X] I have added/updated functional tests where appropriate
* [X] The code follows our [coding standards](https://github.com/companieshouse/styleguides/blob/master/java.md)

__An exhaustive list of peer review checks can be read [here](https://github.com/companieshouse/styleguides/blob/master/java_review.md#developer-actions-prior-to-code-commitreview-started)__

### CODE REVIEW

File: src/main/java/uk/gov/ch/oauth/tokens/SessionSignInModifier.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        SignInInfoMap signInInfo = new SignInInfoMap();
+        signInInfo.setUserProfile(userProfileResponse);
+        signInInfo.setAccessToken(accessToken);
+        signInInfo.setSignedIn(true);
+
+        alterSessionData(session.getData(), signInInfo.toMap());
+    }
+
+    public void alterSessionData(Map<String, Object> currentSessionData,
+            Map<String, Object> desiredSignInInfo) {
```
Comment (Zain-Abbas1) [inline src/main/java/uk/gov/ch/oauth/tokens/SessionSignInModifier.java:31]: Does the method name have to be the same? It's confusing

File: src/main/java/uk/gov/ch/oauth/tokens/SessionSignInModifier.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    // This is necessary as the original data on the session is untyped, but expected to be of the correct types
+    static Map<String, Object> updateSignIn(final Object sInf, final Object sio) {
+        final Map<String, Object> original = (Map<String, Object>) sInf;
+        final Map<String, Object> extras = (Map<String, Object>) sio;
+        original.putAll(extras);
+        return original;
+    }
+
+    public void alterSessionData(Session session, AccessToken accessToken,
+            UserProfile userProfileResponse) {
```
Comment (Zain-Abbas1) [inline src/main/java/uk/gov/ch/oauth/tokens/SessionSignInModifier.java:21]: Why are we passing in session, when we're not using it in this method. I think we can do without.

File: src/main/java/uk/gov/ch/oauth/tokens/SessionSignInModifier.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        signInInfo.setUserProfile(userProfileResponse);
+        signInInfo.setAccessToken(accessToken);
+        signInInfo.setSignedIn(true);
+
+        alterSessionData(session.getData(), signInInfo.toMap());
+    }
+
+    public void alterSessionData(Map<String, Object> currentSessionData,
+            Map<String, Object> desiredSignInInfo) {
+
```
Comment (Zain-Abbas1) [inline src/main/java/uk/gov/ch/oauth/tokens/SessionSignInModifier.java:32]: Can get session data directly in this method instead of passing in `session.getData()`

File: src/main/java/uk/gov/ch/oauth/tokens/SessionSignInModifier.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+public class SessionSignInModifier {
+
+    @SuppressWarnings("unchecked")
+    // This is necessary as the original data on the session is untyped, but expected to be of the correct types
+    static Map<String, Object> updateSignIn(final Object sInf, final Object sio) {
+        final Map<String, Object> original = (Map<String, Object>) sInf;
+        final Map<String, Object> extras = (Map<String, Object>) sio;
+        original.putAll(extras);
+        return original;
+    }
```
Comment (Zain-Abbas1) [inline src/main/java/uk/gov/ch/oauth/tokens/SessionSignInModifier.java:18]: The original method from Oauth2 can be removed now

File: src/test/java/uk/gov/ch/oauth/tokens/OAuthTokenTest.java [hunk 5]
```diff
@@ -14,23 +22,75 @@
...
+        String input = "{\"access_token\":\"TOKEN\",\"expires_in\":1,\"refresh_token\":\"REFRESH_TOKEN\",\"token_type\":\"TOKEN_TYPE\"}";
+        OAuthToken parsedToken = mapper.readValue(input, OAuthToken.class);
+        assertEquals(token.getTokenType(), parsedToken.getTokenType());
+        assertEquals(token.getToken(), parsedToken.getToken());
+        assertEquals(token.getRefreshToken(), parsedToken.getRefreshToken());
+        assertEquals(token.getExpiresIn(), parsedToken.getExpiresIn());
     }
+
+    private void checkNodeValue(Entry<String, JsonNode> entry, OAuthToken token) {
+        System.out.println(entry.getKey());
```
Comment (Zain-Abbas1) [inline src/test/java/uk/gov/ch/oauth/tokens/OAuthTokenTest.java:65]: Delete sysout

File: src/test/java/uk/gov/ch/oauth/tokens/OAuthTokenTest.java [hunk 6]
```diff
@@ -14,23 +22,75 @@
...
+    }
 
-        assertEquals(EXPIRES_IN, data.get(SessionKeys.EXPIRES_IN.getKey()));
-        assertEquals(REFRESH_TOKEN, data.get(SessionKeys.REFRESH_TOKEN.getKey()));
-        assertEquals(TOKEN, data.get(SessionKeys.ACCESS_TOKEN.getKey()));
-        assertEquals(TOKEN_TYPE, data.get(SessionKeys.TOKEN_TYPE.getKey()));
+    @Test
+    @DisplayName("Check AccessToken deserialises correctly.")
+    void setAccessTokenDeserialises() throws JsonProcessingException {
+        String input = "{\"access_token\":\"TOKEN\",\"expires_in\":1,\"refresh_token\":\"REFRESH_TOKEN\",\"token_type\":\"TOKEN_TYPE\"}";
```
Comment (Zain-Abbas1) [inline src/test/java/uk/gov/ch/oauth/tokens/OAuthTokenTest.java:56]: Use `EXPIRES_IN` instead of 1

File: src/test/java/uk/gov/ch/oauth/tokens/SessionSignInModifierTest.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+    private Map<String, Object> mockUserResponseMap;
+
+    @Test
+    void alterSessionData_withObjects_delegatesWithCorrectMaps() {
+        when(mockSession.getData()).thenReturn(sessionMap);
+        when(mockOAuthToken.getAccessTokenAsMap()).thenReturn(mockOAuthTokenMap);
+        when(mockUserResponse.getUserProfileAsMap()).thenReturn(mockUserResponseMap);
+
+        //Guys want a second opinion on whether this is good or bad practice
+        doAnswer(assertNewMapIsCorrect()).when(sessionSignInModifier)
```
Comment (Zain-Abbas1) [inline src/test/java/uk/gov/ch/oauth/tokens/SessionSignInModifierTest.java:50]: This is confusing and hard to follow. What is the purpose of this?

File: src/main/java/uk/gov/ch/oauth/tokens/SessionSignInModifier.java (hunk 1 above)
Comment (CH-JonathanBlake) [reply src/main/java/uk/gov/ch/oauth/tokens/SessionSignInModifier.java:31]: Yes thats what overloading does. Two methods that take different arguments but give the same result.

File: src/main/java/uk/gov/ch/oauth/tokens/SessionSignInModifier.java (hunk 2 above)
Comment (CH-JonathanBlake) [reply src/main/java/uk/gov/ch/oauth/tokens/SessionSignInModifier.java:21]: I chose to pass in session here because it simplifies code on the calling side and it clearly separates the overloaded methods into ones that are interacting with the objects and ones that interact with the maps that represent data.

File: src/main/java/uk/gov/ch/oauth/tokens/SessionSignInModifier.java (hunk 3 above)
Comment (CH-JonathanBlake) [reply src/main/java/uk/gov/ch/oauth/tokens/SessionSignInModifier.java:32]: You're complaining about me passing in Session into the method that calls this. Could you please layout what you think the signatures should look like?

File: src/test/java/uk/gov/ch/oauth/tokens/UserProfileResponseTest.java [hunk 8]
```diff
@@ -34,22 +41,75 @@ void setup() {
...
+        assertEquals(userProfileResponse.getLocale(), parsedToken.getLocale());
+        assertEquals(userProfileResponse.getId(), parsedToken.getId());
+        assertEquals(userProfileResponse.getForename(), parsedToken.getForename());
+        assertEquals(userProfileResponse.getSurname(), parsedToken.getSurname());
+        assertEquals(userProfileResponse.getEmail(), parsedToken.getEmail());
+    }
+
+    private void checkNodeValue(Entry<String, JsonNode> entry,
+            UserProfileResponse userProfileResponse) {
+        System.out.println(entry.getKey());
```
Comment (Zain-Abbas1) [inline src/test/java/uk/gov/ch/oauth/tokens/UserProfileResponseTest.java:76]: Delete sysout please

File: src/main/java/uk/gov/ch/oauth/tokens/SessionSignInModifier.java (hunk 4 above)
Comment (CH-JonathanBlake) [reply src/main/java/uk/gov/ch/oauth/tokens/SessionSignInModifier.java:18]: Good point

File: src/test/java/uk/gov/ch/oauth/tokens/OAuthTokenTest.java (hunk 5 above)
Comment (CH-JonathanBlake) [reply src/test/java/uk/gov/ch/oauth/tokens/OAuthTokenTest.java:65]: Sorry I thought I had all of them

File: src/test/java/uk/gov/ch/oauth/tokens/OAuthTokenTest.java (hunk 6 above)
Comment (CH-JonathanBlake) [reply src/test/java/uk/gov/ch/oauth/tokens/OAuthTokenTest.java:56]: you mean it should look like?
`\"expires_in\":\"EXPIRES_IN\",`?

File: src/test/java/uk/gov/ch/oauth/tokens/SessionSignInModifierTest.java (hunk 7 above)
Comment (CH-JonathanBlake) [reply src/test/java/uk/gov/ch/oauth/tokens/SessionSignInModifierTest.java:50]: Might be better to call.
As this method doesn't return anything and generates a new object which needs appropriate values, this catches any time that there is a call that matches this address and checks that the second argument (newMap) is correct.

File: src/main/java/uk/gov/ch/oauth/tokens/SessionSignInModifier.java (hunk 3 above)
Comment (Zain-Abbas1) [reply src/main/java/uk/gov/ch/oauth/tokens/SessionSignInModifier.java:32]: Please see my comment above, we're getting the data from request attributes, I don't see the benefit of passing in the session when we can remove it from the method signature altogether and get the session within the method. Much like how it was done prior to your change.

```suggestion
public void alterSessionData(Map<String, Object> desiredSignInInfo) {
final Map<String, Object> sData = sessionFactory.getSessionDataFromContext();
```

File: src/main/java/uk/gov/ch/oauth/tokens/SessionSignInModifier.java (hunk 1 above)
Comment (Zain-Abbas1) [reply src/main/java/uk/gov/ch/oauth/tokens/SessionSignInModifier.java:31]: I'm quite familiar with what overloading does, my point here is whether it's needed. There is no need for us to overload the method, in fact even for us to expose this method. As we would always have to set the `SignInInfoMap` data first before we go onto update the map.

File: src/main/java/uk/gov/ch/oauth/Oauth2.java [hunk 9]
```diff
@@ -227,6 +220,7 @@ public void invalidateSession(Session chSession) {
         }
     }
 
+    @SuppressWarnings("unchecked")
```
Comment (ch-rfishwick) [inline src/main/java/uk/gov/ch/oauth/Oauth2.java:223]: Not saying this is wrong, just curious, what is the warning here?

Comment (CH-JonathanBlake) [reply src/main/java/uk/gov/ch/oauth/Oauth2.java:223]: casting the Object into a Map<String,Object>. It could throw errors if the Object isn't actually a Map.
